File: projectBlueWater/src/main/java/com/lasthopesoftware/bluewater/client/playback/file/ExoPlayerPlaybackHandler.java
package com.lasthopesoftware.bluewater.client.playback.file;

import com.google.android.exoplayer2.ExoPlaybackException;
import com.google.android.exoplayer2.PlaybackParameters;
import com.google.android.exoplayer2.Player;
import com.google.android.exoplayer2.SimpleExoPlayer;
import com.google.android.exoplayer2.Timeline;
import com.google.android.exoplayer2.source.TrackGroupArray;
import com.google.android.exoplayer2.trackselection.TrackSelectionArray;
import com.lasthopesoftware.messenger.Messenger;
import com.lasthopesoftware.messenger.promises.MessengerOperator;
import com.lasthopesoftware.messenger.promises.Promise;

import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

import java.io.IOException;

public class ExoPlayerPlaybackHandler
implements
	IPlaybackHandler,
	Player.EventListener,
	MessengerOperator<IPlaybackHandler>,
	Runnable
{
	private static final Logger logger = LoggerFactory.getLogger(ExoPlayerPlaybackHandler.class);

	private final SimpleExoPlayer exoPlayer;
	private final Promise<IPlaybackHandler> playbackHandlerPromise;
	private Messenger<IPlaybackHandler> playbackHandlerMessenger;
	private boolean isPlaying;

	public ExoPlayerPlaybackHandler(SimpleExoPlayer exoPlayer) {
		this.exoPlayer = exoPlayer;
		exoPlayer.addListener(this);

		this.playbackHandlerPromise = new Promise<>((MessengerOperator<IPlaybackHandler>) this);
	}

	@Override
	public boolean isPlaying() {
		return isPlaying;
	}

	@Override
	public void pause() {
		exoPlayer.stop();
		isPlaying = false;
	}

	@Override
	public void setVolume(float volume) {
		exoPlayer.setVolume(volume);
	}

	@Override
	public float getVolume() {
		return exoPlayer.getVolume();
	}

	@Override
	public long getCurrentPosition() {
		return exoPlayer.getCurrentPosition();
	}

	@Override
	public long getDuration() {
		return exoPlayer.getDuration();
	}

	@Override
	public Promise<IPlaybackHandler> promisePlayback() {
		exoPlayer.setPlayWhenReady(true);
		isPlaying = true;
		return playbackHandlerPromise;
	}

	@Override
	public void close() throws IOException {
		isPlaying = false;
		exoPlayer.release();
	}

	@Override
	public void send(Messenger<IPlaybackHandler> playbackHandlerMessenger) {
		this.playbackHandlerMessenger = playbackHandlerMessenger;

		playbackHandlerMessenger.cancellationRequested(this);
	}

	@Override
	public void onTimelineChanged(Timeline timeline, Object manifest) {
		logger.debug("Timeline changed");
	}

	@Override
	public void onTracksChanged(TrackGroupArray trackGroups, TrackSelectionArray trackSelections) {
		logger.debug("Tracks changed");
	}

	@Override
	public void onLoadingChanged(boolean isLoading) {
		logger.debug("Loading changed");
	}

	@Override
	public void onPlayerStateChanged(boolean playWhenReady, int playbackState) {
		logger.debug("Playback state has changed to " + playbackState);

		if (isPlaying && playbackState == Player.STATE_IDLE) {
			logger.warn("The player was playing, but it transitioned to idle! Restarting the player...");
			pause();
			exoPlayer.setPlayWhenReady(true);
		}

		if (playbackState != Player.STATE_ENDED) return;

		isPlaying = false;

		playbackHandlerMessenger.sendResolution(this);
		exoPlayer.removeListener(this);
	}

	@Override
	public void onRepeatModeChanged(int repeatMode) {
		logger.debug("Repeat mode has changed");
	}

	@Override
	public void onPlayerError(ExoPlaybackException error) {
		logger.error("A player error has occurred", error);
		playbackHandlerMessenger.sendRejection(error);
	}

	@Override
	public void onPositionDiscontinuity() {
		logger.debug("Position discontinuity has occurred");
	}

	@Override
	public void onPlaybackParametersChanged(PlaybackParameters playbackParameters) {
		logger.debug("Playback parameters have changed");
	}

	@Override
	public void run() {
		exoPlayer.setPlayWhenReady(false);
		exoPlayer.stop();
		exoPlayer.release();

		isPlaying = false;
	}
}

File: projectBlueWater/src/main/java/com/lasthopesoftware/bluewater/client/playback/file/exoplayer/ExoPlayerPlaybackHandler.kt
package com.lasthopesoftware.bluewater.client.playback.file.exoplayer

import com.google.android.exoplayer2.ExoPlaybackException
import com.google.android.exoplayer2.PlaybackParameters
import com.google.android.exoplayer2.Player
import com.google.android.exoplayer2.Timeline
import com.google.android.exoplayer2.source.TrackGroupArray
import com.google.android.exoplayer2.trackselection.TrackSelectionArray
import com.lasthopesoftware.bluewater.client.playback.exoplayer.PromisingExoPlayer
import com.lasthopesoftware.bluewater.client.playback.file.PlayableFile
import com.lasthopesoftware.bluewater.client.playback.file.PlayedFile
import com.lasthopesoftware.bluewater.client.playback.file.PlayingFile
import com.lasthopesoftware.bluewater.client.playback.file.exoplayer.ExoPlayerPlaybackHandler
import com.lasthopesoftware.bluewater.client.playback.file.exoplayer.playback.PromisedPlayedExoPlayer
import com.lasthopesoftware.bluewater.client.playback.file.exoplayer.progress.ExoPlayerFileProgressReader
import com.lasthopesoftware.bluewater.shared.promises.extensions.ProgressedPromise
import com.namehillsoftware.handoff.promises.Promise
import com.namehillsoftware.lazyj.AbstractSynchronousLazy
import com.namehillsoftware.lazyj.CreateAndHold
import org.joda.time.Duration
import org.slf4j.LoggerFactory

class ExoPlayerPlaybackHandler(private val exoPlayer: PromisingExoPlayer) : PlayableFile, PlayingFile, Player.EventListener, Runnable {

	companion object {
		private val logger = LoggerFactory.getLogger(ExoPlayerPlaybackHandler::class.java)
	}

	private val lazyFileProgressReader: CreateAndHold<ExoPlayerFileProgressReader> = object : AbstractSynchronousLazy<ExoPlayerFileProgressReader>() {
		override fun create(): ExoPlayerFileProgressReader {
			return ExoPlayerFileProgressReader(exoPlayer)
		}
	}

	private val exoPlayerPositionSource: CreateAndHold<ProgressedPromise<Duration, PlayedFile>> = object : AbstractSynchronousLazy<ProgressedPromise<Duration, PlayedFile>>() {
		override fun create(): ProgressedPromise<Duration, PlayedFile> {
			return PromisedPlayedExoPlayer(
				exoPlayer,
				lazyFileProgressReader.getObject(),
				this@ExoPlayerPlaybackHandler)
		}
	}

	var isPlaying = false
		private set

	private var backingDuration = Duration.ZERO

	init {
		exoPlayer.addListener(this)
	}

	private fun pause(): Promise<PromisingExoPlayer> {
		isPlaying = false
		return exoPlayer.setPlayWhenReady(false)
	}

	override fun promisePause(): Promise<PlayableFile> {
		return pause().then { this }
	}

	override fun promisePlayedFile(): ProgressedPromise<Duration, PlayedFile> {
		return exoPlayerPositionSource.getObject()
	}

	override fun getProgress(): Duration {
		return lazyFileProgressReader.getObject().progress
	}

	override val duration: Promise<Duration>
		get() {
			return exoPlayer.getDuration().then { newDuration ->
				if (newDuration == backingDuration.millis) backingDuration
				else Duration.millis(newDuration).also { backingDuration = it }
			}
		}

	override fun promisePlayback(): Promise<PlayingFile> {
		isPlaying = true
		exoPlayer.setPlayWhenReady(true)
		return Promise(this)
	}

	override fun close() {
		isPlaying = false
		exoPlayer.setPlayWhenReady(false)
		exoPlayer.stop()
		exoPlayer.release()
	}

	override fun onTimelineChanged(timeline: Timeline, manifest: Any?, reason: Int) {
		logger.debug("Timeline changed")
	}

	override fun onTracksChanged(trackGroups: TrackGroupArray, trackSelections: TrackSelectionArray) {
		logger.debug("Tracks changed")
	}

	override fun onLoadingChanged(isLoading: Boolean) {
		logger.debug("Loading changed to $isLoading")
	}

	override fun onPlayerStateChanged(playWhenReady: Boolean, playbackState: Int) {
		logger.debug("Playback state has changed to $playbackState")
		if (playbackState != Player.STATE_ENDED) return
		isPlaying = false
		exoPlayer.removeListener(this)
	}

	override fun onRepeatModeChanged(repeatMode: Int) {
		logger.debug("Repeat mode has changed")
	}

	override fun onShuffleModeEnabledChanged(shuffleModeEnabled: Boolean) {}
	override fun onPlayerError(error: ExoPlaybackException) {}
	override fun onPositionDiscontinuity(reason: Int) {}
	override fun onPlaybackParametersChanged(playbackParameters: PlaybackParameters) {
		logger.debug("Playback parameters have changed")
	}

	override fun onSeekProcessed() {}
	override fun run() {
		close()
	}
}
